**#solutions of tutorial problems**

#*solution of problem 1.*

In [40]:
!pip install -q pyomo


In [41]:
from pyomo.environ import * 

this is my problem description\
$\min \ 2x_1 + 15x_2 - 4x_3 - 6x_4 + 7x_5 - 9x_6 - 1.6x_7\\
\text{ s.t } x_1+x_2+x_3+x_4+x_5+x_6+x_7 = 200,\\
x_1 + x_2 ≥ 10,\\
4x_3 + 2x_6 + 3x_7 ≤ 150\\
10x_1 - x_6 + 3x_5 ≥ 20,\\
 x_3- x_5 + 3x_7 ≥ 10,\
 x_1 ≥ 0, x_2 ≥ 12, x_3 ≥ −10, x_4 ≥ 8,\\
 x_5 ≥ 9, x_6 ≥ −1, x_7 ≥ 0$
 
---

In [42]:
#creating model
model = ConcreteModel()

In [43]:
# defing decision variables
model.x1=Var(domain=Reals)
model.x2=Var(domain=Reals)
model.x3=Var(domain=Reals)
model.x4=Var(domain=Reals)
model.x5=Var(domain=Reals)
model.x6=Var(domain=Reals)
model.x7=Var(domain=Reals)

In [44]:
#this is my objective function
model.cost=Objective(expr=2*model.x1 + 15*model.x2 - 4*model.x3 - 6*model.x4 + 7*model.x5 - 9*model.x6 - 1.6*model.x7, sense=minimize)


In [45]:
#these are the constraints.
model.cons1=Constraint(expr=model.x1 + model.x2 + model.x3 + model.x4 + model.x5 + model.x6 + model.x7==200)
model.cons2=Constraint(expr=model.x1 + model.x2 >=10)
model.cons3=Constraint(expr=4*model.x3 + 2*model.x6 + 3*model.x7 <=150)
model.cons4=Constraint(expr=10*model.x1-model.x6+3*model.x5 >=20)
model.cons5=Constraint(expr=model.x1 >=0)
model.cons6=Constraint(expr=model.x2 >=12)
model.cons7=Constraint(expr=model.x3 >=-10)
model.cons8=Constraint(expr=model.x4 >=8)
model.cons9=Constraint(expr=model.x5 >=9)
model.cons10=Constraint(expr=model.x6 >=-1)
model.cons11=Constraint(expr=model.x7 >=0)


In [39]:
model.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   8.8 :  None : False : False :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  12.0 :  None : False : False :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None : -10.0 :  None : False : False :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  85.2 :  None : False : False :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   9.0 :  None : False : False :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  95.0 :  None : False : False :  Reals
    x7 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fi

In [8]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 144793 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [9]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -1065.6
  Upper bound: -1065.6
  Number of objectives: 1
  Number of constraints: 12
  Number of variables: 8
  Number of nonzeros: 23
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.010738134384155273
# ----------------------------------------------------------
#   Solution Information
# ------------------------------

In [10]:
# display solution
print('\nCost = ', model.cost())

print('\nDecision Variables')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())
print('x6 = ', model.x6())
print('x7 = ', model.x7())


print('\nConstraints')
print('Constraint 1  = ', model.cons1())
if model.cons1.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 2  = ', model.cons2())
if model.cons2.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 3  = ', model.cons3())
if model.cons3.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 4  = ', model.cons4())
if model.cons3.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 5  = ', model.cons5())
print('Constraint 6  = ', model.cons6())
print('Constraint 7  = ', model.cons7())
print('Constraint 8 = ', model.cons8())
print('Constraint 9  = ', model.cons9())
print('Constraint 10  = ', model.cons10())
print('Constraint 11 = ', model.cons11())





Cost =  -1065.6

Decision Variables
x1 =  8.8
x2 =  12.0
x3 =  -10.0
x4 =  85.2
x5 =  9.0
x6 =  95.0
x7 =  0.0

Constraints
Constraint 1  =  200.0
constraint is satisfied
Constraint 2  =  20.8
constraint is satisfied
Constraint 3  =  150.0
constraint is satisfied
Constraint 4  =  20.0
constraint is satisfied
Constraint 5  =  8.8
Constraint 6  =  12.0
Constraint 7  =  -10.0
Constraint 8 =  85.2
Constraint 9  =  9.0
Constraint 10  =  95.0
Constraint 11 =  0.0


**3[R]**\
**At optimised condition i obtained following values**

value of objective function at optimum condition\
Cost =  -1065.6

Value of Decision Variables at optimum condition \
x1 =  8.8\
x2 =  12.0\
x3 =  -10.0\
x4 =  85.2\
x5 =  9.0\
x6 =  95.0\
x7 =  0.0\



**4[R]**\
 Value of left side of Constraints at optimum condition \
Constraint 1  =  200.0\
Constraint 2  =  20.8\
Constraint 3  =  150.0\
Constraint 4  =  20.0\
all constraints satisfied.

# here is the solution for second problem of tutorial sheet

In [11]:
!pip install -q pyomo

after reformulating the given equations it will look like it.

$\max 0.043x_1 + 0.027x_2 + 0.025x_3 + 0.022x_4 + 0.045x_5 - 0.0275_y\\
\text{ s.t.   } x_1 + x_2 + x_3 + x_4 + x_5 - y ≤ 100\\
x_2 + x_3 + x_4 ≥ 60\\
.6x_1 + .6x_2 -.4x_4 + 3.6x_5\leq0\\
4x_1 +10x_2 -x_3 + 3x_4 - 3x_5\leq0\\
x_1\geq0 ,x_2\geq0,x_3\geq0,x_4\geq0,x_4\geq0,x_5\geq0$



In [12]:
from pyomo.environ import *

In [13]:
model2=ConcreteModel()

In [14]:

model2.x1=Var(domain=Reals)
model2.x2=Var(domain=Reals)
model2.x3=Var(domain=Reals)
model2.x4=Var(domain=Reals)
model2.x4=Var(domain=Reals)
model2.x5=Var(domain=Reals)
model2.y=Var(domain=Reals,bounds=(0,5))

    'pyomo.core.base.var.SimpleVar'>) on block unknown with a new Component
    (type=<class 'pyomo.core.base.var.SimpleVar'>). This is usually indicative
    of a modelling error. To avoid this warning, use block.del_component() and
    block.add_component().


In [15]:
#here is my objective function
model2.cost=Objective(expr=.043*model2.x1 + .027*model2.x2 + .025*model2.x3 + .022*model2.x4 + .045*model2.x5-.0275*model2.y , sense=maximize)

In [16]:
#here i am defining constraints
model2.cons1=Constraint(expr=model2.x1 + model2.x2 + model2.x3 + model2.x4 + model2.x5 - model2.y <=100)
model2.cons2=Constraint(expr=model2.x2 + model2.x3 + model2.x4 >=60)
model2.cons3=Constraint(expr=.6*model2.x1 + .6*model2.x2 - .4*model2.x3 -.4*model2.x4 + 3.6*model2.x5 <=0)
model2.cons4=Constraint(expr=4*model2.x1 + 10*model2.x2 - model2.x3 + 3*model2.x4 -3*model2.x5  <=0)
model2.cons5=Constraint(expr= model2.x1>=0)
model2.cons6=Constraint(expr= model2.x2>=0)
model2.cons7=Constraint(expr= model2.x3>=0)
model2.cons8=Constraint(expr= model2.x4>=0)
model2.cons9=Constraint(expr= model2.x5>=0)
                      

In [17]:
model2.pprint()

6 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :     5 : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
 

In [18]:
!apt-get install -y -qq glpk-utils

In [19]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model2).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 2.99531818181818
  Upper bound: 2.99531818181818
  Number of objectives: 1
  Number of constraints: 10
  Number of variables: 7
  Number of nonzeros: 25
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.015407323837280273
# ----------------------------------------------------------
#   Solution Information
# ------------

In [20]:
 #display solution
print('cost',model2.cost())

print('\nDecision Variables')
print('x1 = ', model2.x1())
print('x2 = ', model2.x2())
print('x3 = ', model2.x3())
print('x4 = ', model2.x4())
print('x5 = ', model2.x5())
print('y = ', model2.y())

print('\nConstraints')
print('Constraint 1  = ', model2.cons1())
if model2.cons1.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 2  = ', model2.cons2())
if model2.cons1.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 3  = ', model2.cons3())
if model2.cons1.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 4  = ', model2.cons4())
if model2.cons1.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")

cost 2.995318181818181

Decision Variables
x1 =  22.9090909090909
x2 =  0.0
x3 =  77.3181818181818
x4 =  0.0
x5 =  4.77272727272727
y =  5.0

Constraints
Constraint 1  =  99.99999999999996
constraint is satisfied
Constraint 2  =  77.3181818181818
constraint is satisfied
Constraint 3  =  -1.0658141036401503e-14
constraint is satisfied
Constraint 4  =  -1.0658141036401503e-14
constraint is satisfied


3[R]
after optimization \
maximum value of objective function =2.9953 is obtained\
value of decision variables at this optimized condition\
x1 =  22.9090909090909\
x2 =  0.0\
x3 =  77.3181818181818\
x4 =  0.0\
x5 =  4.77272727272727\
y =  5.0\
All the constraints equations are satisfied

**# Here is the model for exercise 3** 

#defining problem 
$\min  45*P + 80*Q\\
P+Q\geq60\\1.25*P+.75*Q\leq350\\P\geq45, Q\geq29$ \\
since the company is manufacturing two product P & Q ,there fore i have taken it as a variable whose unit will be in 'kg'.
production cost  per kg P is Rs 45 and Rs 80 for Q per kg.
so total manufacturing cost can be written like above mentioned objective function.
constraints are defined according to problem.
it is assumed that minimum production of P should be 45kg and Q should be 29kg since this is order placed by a  customer, delivery of minimum this much product is must.
also total available time slot for production of both product in month of september is 350hr ,therefore our total machine run time should be less than equal to it. 

In [21]:
!pip install -q pyomo

In [22]:
from pyomo.environ import *

In [23]:
model3=ConcreteModel()

In [24]:
model3.P=Var(domain=Reals)
model3.Q=Var(domain=Reals)

In [25]:
# defining my objective function
model3.cost=Objective(expr= 45*model3.P + 80*model3.Q , sense=minimize )

In [26]:
# defining the constraints 
model3.cons1=Constraint(expr= model3.P + model3.Q >= 160)
model3.cons2=Constraint(expr= 1.25*model3.P + .75*model3.Q <= 350)
model3.cons3=Constraint(expr= model3.P >=45)
model3.cons4=Constraint(expr= model3.Q >=29)



In [27]:
model3.pprint()

2 Var Declarations
    P : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    Q : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 45*P + 80*Q

4 Constraint Declarations
    cons1 : Size=1, Index=None, Active=True
        Key  : Lower : Body  : Upper : Active
        None : 160.0 : P + Q :  +Inf :   True
    cons2 : Size=1, Index=None, Active=True
        Key  : Lower : Body            : Upper : Active
        None :  -Inf : 1.25*P + 0.75*Q : 350.0 :   True
    cons3 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None :  45.0 :    P :  +Inf :   True
    cons4 : Size=1, Index=None, Active=True
        Key  : Lower

In [28]:
!apt-get install -y -qq glpk-utils

In [29]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model3).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 8215.0
  Upper bound: 8215.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.013573408126831055
# ----------------------------------------------------------
#   Solution Information
# ----------------------------------

In [30]:
print('cost', model3.cost())

print('\ndecision variables')
print('P=',model3.P())
print('Q=', model3.Q())
print('\nConstraints')
print('Constraint 1  = ', model2.cons1())
if model3.cons1.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 2  = ', model2.cons2())
if model3.cons2.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 3  = ', model2.cons3())
if model3.cons3.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 4  = ', model2.cons4())
if model3.cons4.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
  



cost 8215.0

decision variables
P= 131.0
Q= 29.0

Constraints
Constraint 1  =  99.99999999999996
constraint is satisfied
Constraint 2  =  77.3181818181818
constraint is satisfied
Constraint 3  =  -1.0658141036401503e-14
constraint is satisfied
Constraint 4  =  -1.0658141036401503e-14
constraint is satisfied


3.R\
**minimum production cost is 8215 rupee.**\
**At optimum condition we should be producing P=131 kg ,Q=29kg**\
**machine utilisation=77.318 hr**\
also all the constraints are satisfied.


#when demand of major customer changes, then i should change the constraints accordingly.

so the modified problem can be written as \
$\min  45*P + 80*Q\\
P+Q\geq60\\1.25*P+.75*Q\leq350\\P\geq140, Q\geq104$ 

here only the last two constraints are modified.Since another customer placed an order of P=95,Q=75 in addition to previous customer who had made an order of P=45 kg Q=29 kg , therefore minimum production of P=140kg and Q=104kg is must.



In [31]:
model4=ConcreteModel()

In [32]:
model4.P=Var(domain=Reals)
model4.Q=Var(domain=Reals)

In [33]:
# defining my objective function
model4.cost=Objective(expr= 45*model4.P + 80*model4.Q , sense=minimize )

In [34]:
# defining the constraints 
model4.cons1=Constraint(expr= model4.P + model4.Q >= 160)
model4.cons2=Constraint(expr= 1.25*model4.P + .75*model4.Q <= 350)
model4.cons3=Constraint(expr= model4.P >=140)
model4.cons4=Constraint(expr= model4.Q >=104)

In [35]:
model3.pprint()

2 Var Declarations
    P : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None : 131.0 :  None : False : False :  Reals
    Q : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  29.0 :  None : False : False :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 45*P + 80*Q

4 Constraint Declarations
    cons1 : Size=1, Index=None, Active=True
        Key  : Lower : Body  : Upper : Active
        None : 160.0 : P + Q :  +Inf :   True
    cons2 : Size=1, Index=None, Active=True
        Key  : Lower : Body            : Upper : Active
        None :  -Inf : 1.25*P + 0.75*Q : 350.0 :   True
    cons3 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None :  45.0 :    P :  +Inf :   True
    cons4 : Size=1, Index=None, Active=True
        Key  : Lower

In [36]:
!apt-get install -y -qq glpk-utils

In [37]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model4).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 14620.0
  Upper bound: 14620.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011109590530395508
# ----------------------------------------------------------
#   Solution Information
# --------------------------------

In [38]:
print('production cost=',model4.cost())
print('\ndecision variables')
print('P=',model4.P())
print('Q=', model4.Q())
print('\nConstraints')
print('Constraint 1  = ', model4.cons1())
if model4.cons1.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 2  = ', model4.cons2())
if model4.cons2.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 3  = ', model4.cons3())
if model4.cons3.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")
print('Constraint 4  = ', model4.cons4())
if model4.cons4.active==True:
  print("constraint is satisfied")
else:
  print("constraint is not satisfied")

production cost= 14620.0

decision variables
P= 140.0
Q= 104.0

Constraints
Constraint 1  =  244.0
constraint is satisfied
Constraint 2  =  253.0
constraint is satisfied
Constraint 3  =  140.0
constraint is satisfied
Constraint 4  =  104.0
constraint is satisfied


At optimum condition 
minimum production cost= 14620
P=140kg and Q=104kg will be produced .\
machine utilisation =253hr\
**# All the constraints are satisfied**. 

from the result of modeling ,i have concluded that keshar industries is  not utilising the availble capacity of machine to its full capacity.in first case machine utilisation was just 77.31 hr but after the company got anaother order its machine utilisation increased to 253 hr. 
my suggestion to keshar company would be to switch its intention to maximum profit instead of minimum cost of production so that it could utilise its machine efficiently.
keshar industries need to work in marketing domain in order to get few more orders.